### PR TITLE
Added OnResponseCompleted to HttpResponse

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue #160 @lodejard @Tratcher @GrabYourPitchforks

### CODE REVIEW

File: src/Microsoft.AspNet.Http.Core/DefaultHttpResponse.cs [hunk 1]
```diff
@@ -6,9 +6,6 @@
 using System.IO;
 using System.Linq;
 using System.Security.Claims;
-using System.Text;
-using System.Threading.Tasks;
-using Microsoft.AspNet.Http;
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Http.Core/DefaultHttpResponse.cs]: Removed unnecessary usings.

File: src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs [hunk 2]
```diff
@@ -155,6 +155,16 @@ void IHttpResponseFeature.OnSendingHeaders(Action<object> callback, object state
             register(callback, state);
         }
 
+        void IHttpResponseFeature.OnResponseCompleted(Action<object> callback, object state)
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs]: You may also want to expose it in the OWIN environment like `OnSendingHeaders`: https://github.com/aspnet/HttpAbstractions/blob/response-feature/src/Microsoft.AspNet.Owin/OwinEnvironment.cs#L64

**Comment by PinpointTownes** [comment]
Have you considered making this new hook async?

Related discussion (concerning OSH): https://github.com/aspnet/Security/issues/76

File: src/Microsoft.AspNet.Owin/OwinConstants.cs [hunk 3]
```diff
@@ -70,6 +70,7 @@ internal static class CommonKeys
             public const string AppName = "host.AppName";
             public const string Capabilities = "server.Capabilities";
             public const string OnSendingHeaders = "server.OnSendingHeaders";
+            public const string OnResponseCompleted = "server.OnResponseCompleted";
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Owin/OwinConstants.cs]: I'm not sure this is a key http://owin.org/spec/spec/CommonKeys.html

File: src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs [hunk 4]
```diff
@@ -155,6 +155,16 @@ void IHttpResponseFeature.OnSendingHeaders(Action<object> callback, object state
...
+        {
+            var register = Prop<Action<Action<object>, object>>(OwinConstants.CommonKeys.OnResponseCompleted);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs]: Feels like it would be better to throw NotSupportedException /cc @Tratcher

Comment (Tratcher) [inline src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs]: Yeah, just leave this out of OWIN for now. We're trying to restrict the adapter to community defined keys for now.

**Comment by lodejard** [comment]
:shipit:

**Comment by davidfowl** [comment]
@ajaybhargavb rebase this

**Comment by ajaybhargavb** [comment]
Rebased.

**Comment by davidfowl** [comment]
:shipit:

**Comment by ajaybhargavb** [comment]
Can you also look at all the related PRs?

File: src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs (hunk 2 above)
Comment (khellang) [inline src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs]: This? https://github.com/aspnet/HttpAbstractions/blob/08ddbe853130fe10bdde6f8b529e5a885e0974be/src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs#L150-L158

Comment (PinpointTownes) [inline src/Microsoft.AspNet.Owin/OwinFeatureCollection.cs]: Nah, the code you pointed supports calling the `server.OnSendingHeaders` callback from ASP.NET 5, but can't inject `server.OnSendingHeaders` in the OWIN environment, which is what the `OwinEnvironment` class does.

Anyway, @Tratcher said they were only using community defined keys.

**Pull request merged**